File: frontend/src/types/customDashboards.ts
/**
 * TypeScript types for Custom Dashboards, Reports, Templates, Datasets, and Shares.
 *
 * Aligned with backend Pydantic schemas in:
 *   backend/src/api/schemas/custom_dashboards.py
 *
 * Phase 3 - Dashboard Builder UI
 */

// =============================================================================
// Dataset Discovery Types
// =============================================================================

export interface ColumnMetadata {
  column_name: string;
  data_type: string;
  description: string;
  is_metric: boolean;
  is_dimension: boolean;
  is_temporal: boolean;
}

export interface Dataset {
  dataset_name: string;
  dataset_id: number;
  schema: string;
  description: string;
  columns: ColumnMetadata[];
}

export interface DatasetListResponse {
  datasets: Dataset[];
  total: number;
  stale: boolean;
  cached_at: string | null;
}

export interface ConfigWarning {
  column_name: string;
  dataset_name: string;
  message: string;
}

export interface ValidateConfigRequest {
  dataset_name: string;
  referenced_columns: string[];
}

export interface ValidateConfigResponse {
  valid: boolean;
  warnings: ConfigWarning[];
}

// =============================================================================
// Chart Types & Configuration (matches backend ChartConfig schema)
// =============================================================================

/** Valid chart types — must match backend VALID_CHART_TYPES */
export type ChartType = 'line' | 'bar' | 'area' | 'pie' | 'kpi' | 'table';

export type Aggregation = 'SUM' | 'AVG' | 'COUNT' | 'MIN' | 'MAX';

export type FilterOperator =
  | '='
  | '!='
  | '>'
  | '<'
  | '>='
  | '<='
  | 'IN'
  | 'NOT IN'
  | 'LIKE';

export type TimeGrain = 'P1D' | 'P1W' | 'P1M' | 'P3M' | 'P1Y';

export interface MetricConfig {
  column: string;
  aggregation: Aggregation;
  label?: string;
  format?: string;
}

export interface ChartFilter {
  column: string;
  operator: FilterOperator;
  value: unknown;
}

export interface DisplayConfig {
  color_scheme?: string;
  show_legend: boolean;
  legend_position?: string;
  axis_label_x?: string;
  axis_label_y?: string;
}

export interface ChartConfig {
  metrics: MetricConfig[];
  dimensions: string[];
  time_range: string;
  time_grain: TimeGrain;
  filters: ChartFilter[];
  display: DisplayConfig;
}

// =============================================================================
// Grid Position (matches backend GridPosition schema — 12-column grid)
// =============================================================================

export interface GridPosition {
  x: number; // 0-11: column start
  y: number; // 0+: row start
  w: number; // 1-12: width in columns (x + w <= 12)
  h: number; // 1-20: height in rows
}

/** Minimum grid dimensions per chart type (from backend validation) */
export const MIN_GRID_DIMENSIONS: Record<ChartType, { w: number; h: number }> = {
  line: { w: 4, h: 3 },
  bar: { w: 4, h: 3 },
  area: { w: 4, h: 3 },
  pie: { w: 3, h: 3 },
  kpi: { w: 3, h: 2 },
  table: { w: 6, h: 4 },
};

export const GRID_COLS = 12;

/** Widget size labels for layout customizer */
export type WidgetSize = 'small' | 'medium' | 'large' | 'full';

/** Mapping from widget size to column span (12-column grid) */
export const SIZE_TO_COLUMNS: Record<WidgetSize, number> = {
  small: 3,
  medium: 6,
  large: 9,
  full: 12,
};

/** Mapping from column span to widget size label */
export const COLUMNS_TO_SIZE: Record<number, WidgetSize> = {
  3: 'small',
  6: 'medium',
  9: 'large',
  12: 'full',
};

// =============================================================================
// Dashboard Filter (dashboard-level filters)
// =============================================================================

export interface DashboardFilter {
  column: string;
  filter_type: 'date_range' | 'select' | 'multi_select';
  default_value?: unknown;
  dataset_names: string[];
}

// =============================================================================
// Chart Preview Types
// =============================================================================

export interface MetricDefinition {
  label: string;
  column?: string;
  aggregate?: string;
  expressionType?: 'SIMPLE' | 'SQL';
}

export interface FilterDefinition {
  column: string;
  operator: string;
  value: string | number | boolean | null;
}

export interface ChartPreviewRequest {
  dataset_name: string;
  metrics: MetricDefinition[];
  dimensions?: string[];
  filters?: FilterDefinition[];
  time_range?: string;
  time_column?: string | null;
  time_grain?: string;
  viz_type?: ChartType;
}

export interface ChartPreviewResponse {
  data: Record<string, unknown>[];
  columns: string[];
  row_count: number;
  truncated: boolean;
  message: string | null;
  query_duration_ms: number | null;
  viz_type: string;
}

// =============================================================================
// Template Types
// =============================================================================

export type TemplateCategory =
  | 'sales'
  | 'marketing'
  | 'customer'
  | 'product'
  | 'operations';

export interface ReportTemplate {
  id: string;
  name: string;
  description: string;
  category: TemplateCategory;
  thumbnail_url: string | null;
  layout_json: Record<string, unknown>;
  reports_json: Record<string, unknown>[];
  required_datasets: string[];
  min_billing_tier: string;
  sort_order: number;
  is_active: boolean;
}

export interface TemplateListResponse {
  templates: ReportTemplate[];
  total: number;
}

export interface TemplateFilters {
  category?: TemplateCategory;
}

// =============================================================================
// Dashboard Types (matches backend DashboardResponse)
// =============================================================================

export type DashboardStatus = 'draft' | 'published' | 'archived';
export type AccessLevel = 'owner' | 'admin' | 'edit' | 'view';

export interface Dashboard {
  id: string;
  name: string;
  description: string | null;
  status: DashboardStatus;
  layout_json: Record<string, unknown>;
  filters_json: DashboardFilter[] | null;
  template_id: string | null;
  is_template_derived: boolean;
  version_number: number;
  reports: Report[];
  access_level: AccessLevel;
  created_by: string;
  created_at: string;
  updated_at: string;
}

export interface DashboardListResponse {
  dashboards: Dashboard[];
  total: number;
  offset: number;
  limit: number;
  has_more: boolean;
}

export interface DashboardCountResponse {
  current_count: number;
  max_count: number | null;
  can_create: boolean;
}

export interface DashboardFilters {
  status?: DashboardStatus;
  limit?: number;
  offset?: number;
}

export interface CreateDashboardRequest {
  name: string;
  description?: string;
  template_id?: string;
}

export interface UpdateDashboardRequest {
  name?: string;
  description?: string;
  layout_json?: Record<string, unknown>;
  filters_json?: DashboardFilter[];
  expected_updated_at?: string; // ISO datetime for optimistic locking
}

// =============================================================================
// Report Types (matches backend ReportResponse)
// =============================================================================

export interface Report {
  id: string;
  dashboard_id: string;
  name: string;
  description: string | null;
  chart_type: ChartType;
  dataset_name: string;
  config_json: ChartConfig;
  position_json: GridPosition;
  sort_order: number;
  created_by: string;
  created_at: string;
  updated_at: string;
  warnings: string[];
}

export interface CreateReportRequest {
  name: string;
  description?: string;
  chart_type: ChartType;
  dataset_name: string;
  config_json: ChartConfig;
  position_json: GridPosition;
}

export interface UpdateReportRequest {
  name?: string;
  description?: string;
  chart_type?: ChartType;
  config_json?: ChartConfig;
  position_json?: GridPosition;
}

export interface ReorderReportsRequest {
  report_ids: string[];
}

// =============================================================================
// Dashboard Shares Types (matches backend ShareResponse)
// =============================================================================

export type SharePermission = 'view' | 'edit' | 'admin';

export interface DashboardShare {
  id: string;
  dashboard_id: string;
  shared_with_user_id: string | null;
  shared_with_role: string | null;
  permission: SharePermission;
  granted_by: string;
  expires_at: string | null;
  is_expired: boolean;
  created_at: string;
  updated_at: string;
}

export interface ShareListResponse {
  shares: DashboardShare[];
  total: number;
}

export interface CreateShareRequest {
  shared_with_user_id?: string;
  shared_with_role?: string;
  permission: SharePermission;
  expires_at?: string;
}

export interface UpdateShareRequest {
  permission?: SharePermission;
  expires_at?: string;
}

// =============================================================================
// Version History Types (matches backend DashboardVersionResponse)
// =============================================================================

export interface DashboardVersion {
  id: string;
  dashboard_id: string;
  version_number: number;
  change_summary: string;
  created_by: string;
  created_at: string;
}

export interface VersionListResponse {
  versions: DashboardVersion[];
  total: number;
}

// =============================================================================
// Audit Trail Types (matches backend AuditEntryResponse)
// =============================================================================

export interface AuditEntry {
  id: string;
  dashboard_id: string;
  action: string;
  actor_id: string;
  details_json: Record<string, unknown> | null;
  created_at: string;
}

export interface AuditListResponse {
  entries: AuditEntry[];
  total: number;
}

// =============================================================================
// Version Snapshot Types (for version preview)
// =============================================================================

export interface VersionSnapshot {
  dashboard: {
    name: string;
    description: string | null;
    layout_json: Record<string, unknown>;
    filters_json: DashboardFilter[] | null;
  };
  reports: Array<{
    id: string;
    name: string;
    description: string | null;
    chart_type: ChartType;
    dataset_name: string;
    config_json: Record<string, unknown>;
    position_json: GridPosition;
    sort_order: number;
  }>;
}

export interface DashboardVersionDetail extends DashboardVersion {
  snapshot_json: VersionSnapshot;
}

// =============================================================================
// Helpers
// =============================================================================

export function getTemplateCategoryLabel(category: TemplateCategory): string {
  const labels: Record<TemplateCategory, string> = {
    sales: 'Sales',
    marketing: 'Marketing',
    customer: 'Customer',
    product: 'Product',
    operations: 'Operations',
  };
  return labels[category] || category;
}

export function getChartTypeLabel(type: ChartType): string {
  const labels: Record<ChartType, string> = {
    line: 'Line Chart',
    bar: 'Bar Chart',
    area: 'Area Chart',
    pie: 'Pie Chart',
    kpi: 'KPI',
    table: 'Table',
  };
  return labels[type] || type;
}

// =============================================================================
// Builder Catalog Taxonomy
// =============================================================================

/**
 * Business-facing widget category taxonomy used by the builder gallery.
 *
 * Note: `uncategorized` is included to support edit-mode hydration fallback for
 * legacy widgets that cannot be mapped confidently.
 */
export type WidgetCategory =
  | 'all'
  | 'roas'
  | 'sales'
  | 'products'
  | 'customers'
  | 'campaigns'
  | 'uncategorized';

export interface WidgetCategoryMeta {
  id: WidgetCategory;
  name: string;
  icon: string;
  description?: string;
}

export const WIDGET_CATEGORY_META: WidgetCategoryMeta[] = [
  { id: 'all', name: 'All', icon: 'LayoutGrid' },
  { id: 'roas', name: 'ROAS & ROI', icon: 'TrendingUp' },
  { id: 'sales', name: 'Sales', icon: 'DollarSign' },
  { id: 'products', name: 'Products', icon: 'Package' },
  { id: 'customers', name: 'Customers', icon: 'Users' },
  { id: 'campaigns', name: 'Campaigns', icon: 'Megaphone' },
  { id: 'uncategorized', name: 'Uncategorized', icon: 'CircleHelp' },
];

export const CHART_TYPE_TO_WIDGET_CATEGORY: Record<ChartType, WidgetCategory> = {
  kpi: 'roas',
  bar: 'sales',
  line: 'campaigns',
  area: 'sales',
  pie: 'products',
  table: 'customers',
};

export function getWidgetCategoryLabel(category: WidgetCategory): string {
  const found = WIDGET_CATEGORY_META.find((meta) => meta.id === category);
  return found?.name ?? category;
}

export function mapChartTypeToWidgetCategory(type: ChartType): WidgetCategory {
  return CHART_TYPE_TO_WIDGET_CATEGORY[type] ?? 'uncategorized';
}


export function mapChartTypeToWidgetCategoryUnsafe(type: string | null | undefined): WidgetCategory {
  switch (type) {
    case 'kpi':
    case 'bar':
    case 'line':
    case 'area':
    case 'pie':
    case 'table':
      return mapChartTypeToWidgetCategory(type);
    default:
      return 'uncategorized';
  }
}

// =============================================================================
// Dashboard Builder Wizard Types
// =============================================================================

export type BuilderStep = 'select' | 'customize' | 'preview';

/**
 * Widget Catalog Item
 *
 * Individual report config extracted from ReportTemplate.reports_json.
 * Used in wizard step 1 (widget gallery) to let users pick pre-configured reports.
 */
export interface WidgetCatalogItem {
  id: string;                           // Unique ID: `${templateId}-${reportIndex}`
  templateId: string;                   // Parent template ID
  name: string;                         // Report name from template
  description: string;                  // Report description
  category: ChartType;                  // Use chart_type as category for filtering
  chart_type: ChartType;                // Chart type (line, bar, kpi, etc.)
  thumbnail_url?: string;               // Optional preview image from template
  default_config: ChartConfig;          // Pre-configured metrics, dimensions, filters
  required_dataset?: string;            // Dataset name from report config

  // Phase 2 taxonomy compatibility fields (non-breaking additions)
  title?: string;                       // Display title alias of `name`
  icon?: string;                        // Optional Lucide icon name
  defaultSize?: WidgetSize;             // Default card/layout size
  businessCategory?: WidgetCategory;    // Business category taxonomy
}

/**
 * Builder Wizard State
 *
 * Tracks the wizard flow state when creating a new dashboard via the guided wizard.
 * The wizard has 3 steps: select widgets → customize layout → preview & save.
 */
export interface BuilderWizardState {
  isWizardMode: boolean;                // True when in wizard creation flow
  currentStep: BuilderStep;             // Current wizard step
  selectedCategory?: ChartType;         // Filter for widget gallery
  selectedBusinessCategory?: WidgetCategory; // Future business-category filter
  selectedWidgets: Report[];            // Temporary widgets (not persisted until save)
  selectedCatalogItems?: WidgetCatalogItem[]; // Optional catalog item tracking
  dashboardName: string;                // Name for new dashboard
  dashboardDescription: string;         // Description for new dashboard
  previewDateRange?: string;            // Selected date range in preview ('7', '30', '90', 'custom')
  saveAsTemplate: boolean;              // Whether to save as template
  isDirty?: boolean;                    // Optional wizard-specific dirty flag
}

File: backend/src/api/schemas/custom_dashboards.py
"""
Pydantic schemas for Custom Dashboards API.

Request and response models with strict validation for layout JSON,
chart configs, and grid positions. Enforces constraints documented
in the edge case analysis.

Phase: Custom Reports & Dashboard Builder
"""

from typing import Optional, List, Any
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator


# =============================================================================
# Grid Position Validation
# =============================================================================

class GridPosition(BaseModel):
    """Grid position for a report widget within the 12-column layout."""

    x: int = Field(..., ge=0, le=11, description="Column start (0-11)")
    y: int = Field(..., ge=0, description="Row start (0+)")
    w: int = Field(..., ge=1, le=12, description="Width in grid columns (1-12)")
    h: int = Field(..., ge=1, le=20, description="Height in grid rows (1-20)")

    @field_validator("w")
    @classmethod
    def width_must_fit_grid(cls, v: int, info) -> int:
        """Ensure widget doesn't extend beyond 12-column grid."""
        x = info.data.get("x", 0)
        if x + v > 12:
            raise ValueError(f"Widget extends beyond grid: x({x}) + w({v}) > 12")
        return v


# =============================================================================
# Chart Configuration Validation
# =============================================================================

class MetricConfig(BaseModel):
    """Configuration for a single metric in a chart."""

    column: str = Field(..., min_length=1, max_length=255, description="Column name from dataset")
    aggregation: str = Field(..., description="Aggregation function: SUM, AVG, COUNT, MIN, MAX")
    label: Optional[str] = Field(None, max_length=255, description="Display label override")
    format: Optional[str] = Field(None, max_length=50, description="Number format (e.g., ',.2f', '$,.0f')")

    @field_validator("aggregation")
    @classmethod
    def valid_aggregation(cls, v: str) -> str:
        allowed = {"SUM", "AVG", "COUNT", "MIN", "MAX"}
        if v.upper() not in allowed:
            raise ValueError(f"Invalid aggregation '{v}'. Must be one of: {allowed}")
        return v.upper()


class ChartFilter(BaseModel):
    """A single filter condition for a chart."""

    column: str = Field(..., min_length=1, max_length=255)
    operator: str = Field(..., description="Filter operator")
    value: Any = Field(..., description="Filter value")

    @field_validator("operator")
    @classmethod
    def valid_operator(cls, v: str) -> str:
        allowed = {"=", "!=", ">", "<", ">=", "<=", "IN", "NOT IN", "LIKE"}
        if v.upper() not in allowed:
            raise ValueError(f"Invalid operator '{v}'. Must be one of: {allowed}")
        return v.upper()


class DisplayConfig(BaseModel):
    """Display settings for a chart."""

    color_scheme: Optional[str] = Field("default", max_length=50)
    show_legend: bool = Field(True)
    legend_position: Optional[str] = Field("bottom")
    axis_label_x: Optional[str] = Field(None, max_length=255)
    axis_label_y: Optional[str] = Field(None, max_length=255)


class ChartConfig(BaseModel):
    """Full chart configuration for a custom report."""

    metrics: List[MetricConfig] = Field(..., min_length=1, max_length=20)
    dimensions: List[str] = Field(default_factory=list, max_length=5)
    time_range: str = Field("Last 30 days", max_length=100)
    time_grain: str = Field("P1D", description="ISO 8601 duration: P1D, P1W, P1M")
    filters: List[ChartFilter] = Field(default_factory=list, max_length=20)
    display: DisplayConfig = Field(default_factory=DisplayConfig)

    @field_validator("time_grain")
    @classmethod
    def valid_time_grain(cls, v: str) -> str:
        allowed = {"P1D", "P1W", "P1M", "P3M", "P1Y"}
        if v not in allowed:
            raise ValueError(f"Invalid time_grain '{v}'. Must be one of: {allowed}")
        return v


# =============================================================================
# Dashboard Filter Schemas
# =============================================================================

class DashboardFilter(BaseModel):
    """Dashboard-level filter definition."""

    column: str = Field(..., min_length=1, max_length=255)
    filter_type: str = Field(..., description="date_range, select, multi_select")
    default_value: Optional[Any] = None
    dataset_names: List[str] = Field(default_factory=list, description="Datasets this filter applies to")


# =============================================================================
# Request Models
# =============================================================================

class CreateDashboardRequest(BaseModel):
    """Request to create a new custom dashboard."""

    name: str = Field(..., min_length=1, max_length=255, description="Dashboard name")
    description: Optional[str] = Field(None, max_length=2000)
    template_id: Optional[str] = Field(None, max_length=36, description="Create from template")

    @field_validator("name")
    @classmethod
    def name_not_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Dashboard name cannot be blank")
        return v.strip()


class UpdateDashboardRequest(BaseModel):
    """Request to update dashboard metadata or layout."""

    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=2000)
    layout_json: Optional[dict] = None
    filters_json: Optional[List[DashboardFilter]] = None
    expected_updated_at: Optional[datetime] = Field(
        None,
        description="Optimistic lock: reject if dashboard.updated_at != this value (409 Conflict)",
    )

    @field_validator("name")
    @classmethod
    def name_not_blank(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and not v.strip():
            raise ValueError("Dashboard name cannot be blank")
        return v.strip() if v else v


class DuplicateDashboardRequest(BaseModel):
    """Request to duplicate an existing dashboard."""

    new_name: str = Field(..., min_length=1, max_length=255)


# =============================================================================
# Report Request Models
# =============================================================================

VALID_CHART_TYPES = {"line", "bar", "area", "pie", "kpi", "table"}


class CreateReportRequest(BaseModel):
    """Request to add a new report/chart to a dashboard."""

    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=2000)
    chart_type: str = Field(..., description="Visualization type")
    dataset_name: str = Field(..., min_length=1, max_length=255)
    config_json: ChartConfig
    position_json: GridPosition

    @field_validator("chart_type")
    @classmethod
    def valid_chart_type(cls, v: str) -> str:
        if v not in VALID_CHART_TYPES:
            raise ValueError(f"Invalid chart_type '{v}'. Must be one of: {VALID_CHART_TYPES}")
        return v

    @field_validator("config_json")
    @classmethod
    def validate_metrics_for_chart_type(cls, v: ChartConfig, info) -> ChartConfig:
        chart_type = info.data.get("chart_type")
        if chart_type == "kpi" and len(v.metrics) != 1:
            raise ValueError("KPI chart requires exactly 1 metric")
        if chart_type == "pie":
            if len(v.metrics) != 1:
                raise ValueError("Pie chart requires exactly 1 metric")
            if len(v.dimensions) != 1:
                raise ValueError("Pie chart requires exactly 1 dimension")
        return v


class UpdateReportRequest(BaseModel):
    """Request to update an existing report's config."""

    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = Field(None, max_length=2000)
    chart_type: Optional[str] = None
    config_json: Optional[ChartConfig] = None
    position_json: Optional[GridPosition] = None

    @field_validator("chart_type")
    @classmethod
    def valid_chart_type(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in VALID_CHART_TYPES:
            raise ValueError(f"Invalid chart_type '{v}'. Must be one of: {VALID_CHART_TYPES}")
        return v


class ReorderReportsRequest(BaseModel):
    """Request to reorder reports within a dashboard."""

    report_ids: List[str] = Field(..., min_length=1, description="Ordered list of report IDs")


# =============================================================================
# Share Request Models
# =============================================================================

VALID_SHARE_PERMISSIONS = {"view", "edit", "admin"}
VALID_SHARE_ROLES = {"merchant_admin", "merchant_viewer", "agency_admin", "agency_viewer"}


class CreateShareRequest(BaseModel):
    """Request to share a dashboard with a user or role."""

    shared_with_user_id: Optional[str] = Field(None, max_length=255)
    shared_with_role: Optional[str] = Field(None, max_length=50)
    permission: str = Field("view", description="Permission level: view, edit, admin")
    expires_at: Optional[datetime] = None

    @field_validator("permission")
    @classmethod
    def valid_permission(cls, v: str) -> str:
        if v not in VALID_SHARE_PERMISSIONS:
            raise ValueError(f"Invalid permission '{v}'. Must be one of: {VALID_SHARE_PERMISSIONS}")
        return v

    @field_validator("shared_with_role")
    @classmethod
    def valid_role(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in VALID_SHARE_ROLES:
            raise ValueError(f"Invalid role '{v}'. Must be one of: {VALID_SHARE_ROLES}")
        return v

    def model_post_init(self, __context: Any) -> None:
        if not self.shared_with_user_id and not self.shared_with_role:
            raise ValueError("Either shared_with_user_id or shared_with_role must be provided")
        if self.shared_with_user_id and self.shared_with_role:
            raise ValueError("Provide shared_with_user_id or shared_with_role, not both")


class UpdateShareRequest(BaseModel):
    """Request to update a share's permission or expiry."""

    permission: Optional[str] = None
    expires_at: Optional[datetime] = None

    @field_validator("permission")
    @classmethod
    def valid_permission(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in VALID_SHARE_PERMISSIONS:
            raise ValueError(f"Invalid permission '{v}'. Must be one of: {VALID_SHARE_PERMISSIONS}")
        return v


# =============================================================================
# Response Models
# =============================================================================

class ReportResponse(BaseModel):
    """Response model for a single custom report."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    dashboard_id: str
    name: str
    description: Optional[str] = None
    chart_type: str
    dataset_name: str
    config_json: dict
    position_json: dict
    sort_order: int
    created_by: str
    created_at: datetime
    updated_at: datetime
    warnings: List[str] = Field(default_factory=list, description="Validation warnings (e.g., missing columns)")


class DashboardResponse(BaseModel):
    """Response model for a single custom dashboard."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    description: Optional[str] = None
    status: str
    layout_json: dict
    filters_json: Optional[list] = None
    template_id: Optional[str] = None
    is_template_derived: bool
    version_number: int
    reports: List[ReportResponse] = Field(default_factory=list)
    access_level: str = Field("owner", description="Caller's access level: owner, admin, edit, view")
    created_by: str
    created_at: datetime
    updated_at: datetime


class DashboardListResponse(BaseModel):
    """Paginated response for listing dashboards."""

    dashboards: List[DashboardResponse]
    total: int
    offset: int
    limit: int
    has_more: bool


class DashboardVersionResponse(BaseModel):
    """Response model for a dashboard version."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    dashboard_id: str
    version_number: int
    change_summary: str
    created_by: str
    created_at: datetime


class DashboardVersionDetailResponse(DashboardVersionResponse):
    """Response model for a single version with its full snapshot."""

    snapshot_json: dict


class VersionListResponse(BaseModel):
    """Paginated response for listing versions."""

    versions: List[DashboardVersionResponse]
    total: int


class ShareResponse(BaseModel):
    """Response model for a dashboard share."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    dashboard_id: str
    shared_with_user_id: Optional[str] = None
    shared_with_role: Optional[str] = None
    permission: str
    granted_by: str
    expires_at: Optional[datetime] = None
    is_expired: bool = False
    created_at: datetime
    updated_at: datetime


class ShareListResponse(BaseModel):
    """Response for listing shares."""

    shares: List[ShareResponse]
    total: int


class AuditEntryResponse(BaseModel):
    """Response model for a dashboard audit entry."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    dashboard_id: str
    action: str
    actor_id: str
    details_json: Optional[dict] = None
    created_at: datetime


class AuditListResponse(BaseModel):
    """Response for listing audit entries."""

    entries: List[AuditEntryResponse]
    total: int


class DashboardCountResponse(BaseModel):
    """Response with dashboard count vs limit for entitlement display."""

    current_count: int
    max_count: Optional[int] = Field(None, description="None means unlimited")
    can_create: bool


# =============================================================================
# Template Response Models
# =============================================================================

class TemplateResponse(BaseModel):
    """Response model for a report template."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    description: str
    category: str
    thumbnail_url: Optional[str] = None
    layout_json: dict
    reports_json: list
    required_datasets: list
    min_billing_tier: str
    sort_order: int
    is_active: bool


class TemplateListResponse(BaseModel):
    """Response for listing templates."""

    templates: List[TemplateResponse]
    total: int


# =============================================================================
# Dataset Response Models
# =============================================================================

class DatasetColumnResponse(BaseModel):
    """Response for a single dataset column."""

    name: str
    type: str = Field(..., description="Data type: numeric, string, datetime, boolean")
    description: Optional[str] = None
    is_metric: bool
    is_dimension: bool


class DatasetResponse(BaseModel):
    """Response for a dataset's metadata."""

    name: str
    description: Optional[str] = None
    schema_name: str
    columns: List[DatasetColumnResponse]


class DatasetListResponse(BaseModel):
    """Response for listing available datasets."""

    datasets: List[DatasetResponse]
    total: int
